pkg/plugins/policies/matchers/match.go

```diff
@@ -1,14 +1,104 @@
 package matchers
 
 import (
+	"sort"
+
+	"github.com/pkg/errors"
+
+	common_proto "github.com/kumahq/kuma/api/common/v1alpha1"
+	mesh_proto "github.com/kumahq/kuma/api/mesh/v1alpha1"
 	core_mesh "github.com/kumahq/kuma/pkg/core/resources/apis/mesh"
-	"github.com/kumahq/kuma/pkg/core/resources/model"
+	core_model "github.com/kumahq/kuma/pkg/core/resources/model"
 	core_xds "github.com/kumahq/kuma/pkg/core/xds"
 	xds_context "github.com/kumahq/kuma/pkg/xds/context"
 )
 
+type ResourceSpecWithTargetRef interface {
+	core_model.ResourceSpec
+	GetTargetRef() *common_proto.TargetRef
+}
+
 // MatchedPolicies match policies using the standard matchers using targetRef (madr-005)
-func MatchedPolicies(model.ResourceType, *core_mesh.DataplaneResource, xds_context.Resources) (core_xds.TypedMatchingPolicies, error) {
-	// TODO @lobkovilya Implement standard matching strategy
-	return core_xds.TypedMatchingPolicies{}, nil
+func MatchedPolicies(rType core_model.ResourceType, dpp *core_mesh.DataplaneResource, resources xds_context.Resources) (core_xds.TypedMatchingPolicies, error) {
+	policies := resources.ListOrEmpty(rType)
+
+	matchedPolicies := []core_model.Resource{}
+	for _, policy := range policies.GetItems() {
+		spec, ok := policy.GetSpec().(ResourceSpecWithTargetRef)
+		if !ok {
+			return core_xds.TypedMatchingPolicies{}, errors.Errorf("resource type %v doesn't support TargetRef matching", rType)
+		}
+
+		targetRef := spec.GetTargetRef()
+		if isDataplaneSelectedByTargetRef(targetRef, dpp) {
+			matchedPolicies = append(matchedPolicies, policy)
+		}
+	}
+
+	sort.Sort(ByTargetRef(matchedPolicies))
+
+	return core_xds.TypedMatchingPolicies{
+		Type:              rType,
+		DataplanePolicies: matchedPolicies,
+	}, nil
+}
+
+func isDataplaneSelectedByTargetRef(tr *common_proto.TargetRef, dpp *core_mesh.DataplaneResource) bool {
+	switch tr.GetKindEnum() {
+	case common_proto.TargetRef_Mesh:
+		return true
+	case common_proto.TargetRef_MeshSubset:
+		return isDataplaneSelectedByTags(tr.GetTags(), dpp)
+	case common_proto.TargetRef_MeshService:
+		return isDataplaneSelectedByTags(map[string]string{
+			mesh_proto.ServiceTag: tr.GetName(),
+		}, dpp)
+	case common_proto.TargetRef_MeshServiceSubset:
+		trTags := tr.GetTags()
+		trTags[mesh_proto.ServiceTag] = tr.GetName()
+		return isDataplaneSelectedByTags(trTags, dpp)
+	default:
+		return false
+	}
+}
+
+func isDataplaneSelectedByTags(tags map[string]string, dpp *core_mesh.DataplaneResource) bool {
+	for _, inbound := range dpp.Spec.GetNetworking().GetInbound() {
+		if isInboundSelectedBy(tags, inbound) {
+			return true
+		}
+	}
+	return false
+}
+
+func isInboundSelectedBy(tags map[string]string, inbound *mesh_proto.Dataplane_Networking_Inbound) bool {
```

Comment:
nit: personally I'd find this more readable if this second function was inlined in the first one

Reply:
I'm not sure, the method will look like:
```go
func isDataplaneSelectedByTags(tags map[string]string, dpp *core_mesh.DataplaneResource) bool {
InboundLoop:
	for _, inbound := range dpp.Spec.GetNetworking().GetInbound() {
		for k, v := range tags {
			if inboundValue, inboundHasTag := inbound.Tags[k]; !inboundHasTag || inboundValue != v {
				continue InboundLoop
			}
		}
		return true
	}
	return false
}
```
I think it's less readable when inlined